### PR TITLE
Fix/dont scare users with humungous thread dumps

### PR DESCRIPTION
**Pull request by gsheasby**
**Save timestamp-related thread dumps to temp files**
Output to unique [temporary files](https://github.com/palantir/atlasdb/files/611207/atlas-timestamps-log553351070509726618.txt). These were 13-25KB for our unit tests, probably larger (but not super large) for an actual application.

Potential improvement: hard-code the file name and replace if the file already exists.

**Synchronise getTimestampService**
Avoid a possible race condition that would prevent us from detecting a failure scenario (when two timestamp services are fetched at once, we previously might not have noticed).

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1275)

### CODE REVIEW

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ServiceDiscoveringAtlasSupplier.java [hunk 1]
```diff
@@ -59,31 +63,68 @@ public KeyValueService getKeyValueService() {
         return keyValueService.get();
     }
 
-    public TimestampService getTimestampService() {
+    public synchronized TimestampService getTimestampService() {
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ServiceDiscoveringAtlasSupplier.java]: 👍 well spotted.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ServiceDiscoveringAtlasSupplier.java [hunk 2]
```diff
@@ -59,31 +63,68 @@ public KeyValueService getKeyValueService() {
...
+    }
+
+    private String saveThreadDumpsToFile(File file) throws IOException {
+        try (FileOutputStream outputStream = new FileOutputStream(file)) {
+            writeStringToStream(outputStream,
+                    "This file contains thread dumps that will be useful for the AtlasDB Dev team, in case you hit a "
+                            + "MultipleRunningTimestampServices error. In this case, please send this file to them.\n");
+            writeStringToStream(outputStream, "First thread dump: " + timestampServiceCreationInfo);
+            writeStringToStream(outputStream, "Second thread dump: " + ThreadDumps.programmaticThreadDump());
+            outputStream.close();
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ServiceDiscoveringAtlasSupplier.java]: Do we need this? I thought try with resources would close the stream for you.

Comment (SerialVelocity) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ServiceDiscoveringAtlasSupplier.java]: You are correct. That's how try with resources works :)

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ServiceDiscoveringAtlasSupplier.java [hunk 3]
```diff
@@ -59,31 +63,68 @@ public KeyValueService getKeyValueService() {
...
+        return saveThreadDumpsToFile(file);
+    }
+
+    private String saveThreadDumpsToFile(File file) throws IOException {
+        try (FileOutputStream outputStream = new FileOutputStream(file)) {
+            writeStringToStream(outputStream,
+                    "This file contains thread dumps that will be useful for the AtlasDB Dev team, in case you hit a "
+                            + "MultipleRunningTimestampServices error. In this case, please send this file to them.\n");
+            writeStringToStream(outputStream, "First thread dump: " + timestampServiceCreationInfo);
+            writeStringToStream(outputStream, "Second thread dump: " + ThreadDumps.programmaticThreadDump());
```
Comment (SerialVelocity) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ServiceDiscoveringAtlasSupplier.java]: Should these have a `\n` on the end?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ServiceDiscoveringAtlasSupplier.java (hunk 2 above)
Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ServiceDiscoveringAtlasSupplier.java]: Yep - at first this was not inside a try-with-resources. Findbugs complained. Will remove this line.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ServiceDiscoveringAtlasSupplier.java (hunk 3 above)
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ServiceDiscoveringAtlasSupplier.java]: The thread dumps end with quite a lot of space, so the output is definitely readable as is; I think it's fine either way.

**Comment by jboreiko** [comment]
An issue was brought up today regarding a concern about filing up the temp directory.  Are we able to instead overwrite the output file so at any given time we have at most 0 or 1 instances of the thread dumps?

**Pull request merged**